sidekiq/queue.go

```diff
@@ -94,28 +94,75 @@ func NewQueue(client redis.UniversalClient) Queue {
 	// https://github.com/mperham/sidekiq/blob/455e9d56f46f0299eaf3b761596207e15f906a39/lib/sidekiq/fetch.rb#L37
 	// This improves OSS version of sidekiq and will not lose sidekiq jobs.
 	dequeueScript := redis.NewScript(`
-	local sidekiq_ns = ARGV[1]
-	local sidekiq_queue = ARGV[2]
-
-	local queue_key = table.concat({"queue", sidekiq_queue}, ":")
-	if sidekiq_ns ~= "" then
-		queue_key = table.concat({sidekiq_ns, queue_key}, ":")
-	end
+	local queue_ns = ARGV[1]
+	local queue_id = ARGV[2]
 
+	local queue_key = table.concat({queue_ns, queue_id}, ":")
 	return redis.call("lindex", queue_key, -1)
 	`)
 
 	ackScript := redis.NewScript(`
+	local queue_ns = ARGV[1]
+	local queue_id = ARGV[2]
+	local jobm = ARGV[3]
+
+	local queue_key = table.concat({queue_ns, queue_id}, ":")
+	return redis.call("lrem", queue_key, -1, jobm)
+	`)
+
+	dequeueStartScript := redis.NewScript(`
 	local sidekiq_ns = ARGV[1]
 	local sidekiq_queue = ARGV[2]
-	local jobm = ARGV[3]
+	local queue_ns = ARGV[3]
+	local queue_id = ARGV[4]
+	local at = tonumber(ARGV[5])
+	local expire_in_sec = tonumber(ARGV[6])
 
 	local queue_key = table.concat({"queue", sidekiq_queue}, ":")
 	if sidekiq_ns ~= "" then
 		queue_key = table.concat({sidekiq_ns, queue_key}, ":")
 	end
+	local pullers_key = table.concat({queue_ns, "pullers"}, ":")
+	if redis.call("zadd", pullers_key, "nx", at + expire_in_sec, queue_id) == 0 then
+		return 0
+	end
+	local puller_queue_key = table.concat({queue_ns, queue_id}, ":")
+	local old_queue_ids = redis.call("zrangebyscore", pullers_key, "-inf", at, "limit", 0, 1)
+	for i, old_queue_id in pairs(old_queue_ids) do
+		local old_puller_queue_key = table.concat({queue_ns, old_queue_id}, ":")
+		if redis.call("exists", old_puller_queue_key) == 1 then
+			redis.call("rename", old_puller_queue_key, puller_queue_key)
+		end
+		redis.call("zrem", pullers_key, old_queue_id)
+		return 1
```

Comment:
I don't understand this loop -- it always returns.  Just more convenient to get the first id?

Reply:
yes.